operators/pkg/instmetrics/metrics_scraper.go

```diff
@@ -0,0 +1,195 @@
+// Copyright 2020-2022 Politecnico di Torino
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package instmetrics
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	runtimeapi "k8s.io/cri-api/pkg/apis/runtime/v1alpha2"
+
+	clctx "github.com/netgroup-polito/CrownLabs/operators/pkg/context"
+	"github.com/netgroup-polito/CrownLabs/operators/pkg/forge"
+	"github.com/netgroup-polito/CrownLabs/operators/pkg/utils"
+)
+
+// CustomMetrics stores desired metrics of a container.
+type CustomMetrics struct {
+	CPUPerc   float32 `json:"cpu"`
+	MemBytes  uint64  `json:"mem"`
+	DiskBytes uint64  `json:"disk"`
+}
+
+func (c CustomMetrics) String() string {
+	return fmt.Sprintf("[CPU] %v \t[MEM] %v \t[DISK] %v", c.CPUPerc, c.MemBytes, c.DiskBytes)
+}
+
+// MetricsScraper interface.
+type MetricsScraper struct {
+	Log           logr.Logger
+	UpdatePeriod  time.Duration
+	RuntimeClient *RemoteRuntimeServiceClient
+	// Containers ignored while retrieving metrics.
+	ignoredContainerNames []string
+	// Stats used to evaluate resource usages changes
+	// <key=podName, val=ContainerStats>
+	oldStats map[string]*runtimeapi.ContainerStats
+	// Stats used to evaluate resource usages changes
+	// <key=podName, val=CustomMetrics>
+	cachedMetrics      map[string]*CustomMetrics
+	cachedMetricsMutex sync.RWMutex
+}
+
+// Start scraping metrics.
+func (ms *MetricsScraper) Start(ctx context.Context) {
+	ms.ignoredContainerNames = []string{
+		forge.XVncName,
+		forge.WebsockifyName,
+		forge.ContentDownloaderName,
+		forge.ContentUploaderName,
+		forge.MyDriveName,
+	}
+	ms.oldStats = make(map[string]*runtimeapi.ContainerStats)
+
+	// Periodically scrape updated metrics for watched containers.
+	ticker := time.NewTicker(ms.UpdatePeriod)
+	ms.Log.Info("MetricsScraper started", "updatePeriod", ms.UpdatePeriod)
+	for range ticker.C {
+		log := ms.Log.WithValues("timeout", ms.UpdatePeriod)
+		ctx, cancel := context.WithTimeout(ctx, ms.UpdatePeriod)
+
+		if err := ms.fillCachedMetrics(ctx); err != nil {
+			log.Error(err, "Error retrieving containerStats")
+		}
+
+		// Operation running the context completed
+		cancel()
+	}
+}
+
+func (ms *MetricsScraper) fillCachedMetrics(ctx context.Context) error {
+	log := ms.Log.WithName("fill-cached-metrics")
+
+	// Update containerIDs list.
+	containerIDs, err := ms.findApplicationContainerIDs(ctx)
+	if err != nil {
+		log.Error(err, "Error updating containerIDs list")
+	}
```

Comment:
Uhm, I actually just meant that it's not much useful to keep going if you failed the update (I guess?) from within fillCachedMetrics. In that case I'd just return and wait for the next iteration... I'm not sure about the error counter anyway, I'm rather sure the application would break in some other way if retrieval errors occur with gRPC...

Reply:
You are absolutely correct. I was wrong: this piece of code is executed in a loop, so just returning the error and go for the next iteration is the right behavure.